### DISCUSSION TITLE
Does redirect() work in custom middleware?

### DISCUSSION THREAD

**Discussion by mschoeffmann** [livewire/livewire#5359] [1. Help]
Hello everyone,

I have one question - maybe I misunderstood something regarding livewire and middlewares ...

Using a custom middleware for (additional) authentication.

For all non-livewire calls, it works as expected:
The user is redirected to the login page if something is not right, and no further code is executed.

But if the same middleware is called during a livewire update, the redirect is ignored and all the further code (inside the component for example) is executed.

```php
    public function handle(Request $request, Closure $next)
    {
        if (letsAssumeAllIsOkay === true) {
            return $next($request);
        }

        return redirect('/login');
    }
```

It would be fine if the redirect is not possible, but right now also further code in the component is executed as if the middleware did not exist.
Calling `abort(401)` works as a workaround, but why is further code in the component executed without calling the $next closure?
And: Why is the "what happens after return" different from non-livewire requests to livewire requests?

Thanks for you help/advice :-)

**Comment by joshhanley** [comment]
@mschoeffmann Livewire has a custom redirect handler for any subsequent Livewire requests, that takes the url and sends it to the frontend as part of the request, and the front end does the redirect, due to it being an API call (hence native browser redirect won't work).

But I believe that will only work within a Livewire component.

I personally add authentication inside my Livewire component to ensure it all works as expected and then redirects get processed correctly if authentication/authorisation fails.

Hope this helps!

**Reply by c0700** [reply]
Hello @joshhanley, could you share your authentication implementation on Livewire component side ? :)

**Reply by joshhanley** [reply]
@c0700 yep, all I do, is this:

```php
use Illuminate\Foundation\Auth\Access\AuthorizesRequests;

class MyComponent extends Component
{
    public function mount()
    {
        $this->authorize('gate');
    }
}
```

Which returns a 403. So you could just change that to be an `if()` check to see if user has authorisation, and if not, then call `$this->redirect()`.

Hope this helps!

**Reply by c0700** [reply]
Your implementation is not fully safe : the `mount` method is call only at the first request (page loading). But after, for ajax requests (livewire actions), `mount` is not recall. It's safer to check authorizations in `boot`method (cf [lifecycle hooks documentation](https://laravel-livewire.com/docs/2.x/lifecycle-hooks)).

On my side, I search a solution to check some authorizations for all requests, and redirect the user at the page loading if authorizations are wrong, to keep a great user experience. The problem is that redirection doesn't work on ajax requests (= ajax requests are perform even if a redirection is call).

I see 2 solutions. The first is to implement all checks on the Livewire component side :

```php
use Illuminate\Foundation\Auth\Access\AuthorizesRequests;

class MyComponent extends Component
{
    private bool $redirect;

    public function boot()
    {
        $this->redirect = false;
    }

    public function mount()
    {
        if (Gate::inspect('auth')->denied()) {
            $this->redirect = true;
            return redirect()->to('/login');    
        }
    }

    public function booted()
    {
        if (!$this->redirect && Gate::inspect('auth')->denied()) {
            throw new AuthorizationException();
        }
    }
}
```

The second is to have middlewares just for redirection, and to check authorizations in the `boot` method of the Livewire component. To avoid redundance, use a gate in middleware and in `boot` method.

For the moment, I used the second solution.

**Reply by joshhanley** [reply]
@c0700 yep, boot would be more secure than mount! 🙂
